lib/GADS/API.pm

```diff
@@ -546,6 +546,40 @@ any ['get', 'post'] => '/api/:sheet/records' => require_login sub {
     _get_records();
 };
 
+post '/api/settings/logo' => require_login sub {
+    my $site = var 'site';
+
+    if ( my $file = upload('file') ) {
+        #should this not return a json response with error set to 1?
+        forwardHome({ danger => "You do not have permission to manage system settings"}, '')
+            unless logged_in_user->permission->{superadmin};
+
+        my $txn_scope_guard = schema->txn_scope_guard;
+
+        $site->update({ site_logo => $file->content });
+
+        $txn_scope_guard->commit;
+
+        content_type 'application/json';
+        # 201 is created rather than ok
+        status 201;
+        return encode_json(
+            {
+                error => 0,
+                url   => '/settings/logo'
+            }
+        );
+    } else {
```

Comment:
Rather than this `else` statement, you could instead start the block with:
```
my $file = upload('file')
  or error __"No file provided";
```

Reply:
As this is done using AJAX, I thought it more "sensible" to use the JSON as (it seemed to be) intended and include an error response to display to the user?

Reply:
The above statement (using `error`) will automatically render JSON. See https://github.com/ctrlo/GADS/blob/master/lib/GADS/API.pm#L32 - so no need to manually write a JSON return statement :)